File: bindings/swift/Tests/ProvenTests/ProvenTests.swift
// SPDX-License-Identifier: PMPL-1.0
// SPDX-FileCopyrightText: 2025 Hyperpolymath

import XCTest
@testable import Proven

final class ProvenTests: XCTestCase {

    // MARK: - SafeMath Tests

    func testSafeDiv() {
        XCTAssertEqual(SafeMath.div(10, 2), 5)
        XCTAssertNil(SafeMath.div(10, 0))
    }

    func testSafeMod() {
        XCTAssertEqual(SafeMath.mod(10, 3), 1)
        XCTAssertNil(SafeMath.mod(10, 0))
    }

    func testSafeAdd() {
        XCTAssertEqual(SafeMath.add(1, 2), 3)
        XCTAssertNil(SafeMath.add(Int.max, 1))
    }

    func testSafeSub() {
        XCTAssertEqual(SafeMath.sub(5, 3), 2)
        XCTAssertNil(SafeMath.sub(Int.min, 1))
    }

    func testSafeMul() {
        XCTAssertEqual(SafeMath.mul(3, 4), 12)
        XCTAssertNil(SafeMath.mul(Int.max, 2))
    }

    // MARK: - SafeString Tests

    func testEscapeHtml() {
        XCTAssertEqual(SafeString.escapeHtml("<script>"), "&lt;script&gt;")
        XCTAssertEqual(SafeString.escapeHtml("a & b"), "a &amp; b")
        XCTAssertEqual(SafeString.escapeHtml("\"quoted\""), "&quot;quoted&quot;")
    }

    func testEscapeSql() {
        XCTAssertEqual(SafeString.escapeSql("it's"), "it''s")
    }

    func testEscapeJs() {
        XCTAssertEqual(SafeString.escapeJs("line\nbreak"), "line\\nbreak")
        XCTAssertEqual(SafeString.escapeJs("tab\there"), "tab\\there")
    }

    func testTruncateSafe() {
        XCTAssertEqual(SafeString.truncateSafe("hello world", maxLength: 5), "he...")
        XCTAssertEqual(SafeString.truncateSafe("hi", maxLength: 10), "hi")
    }

    // MARK: - SafePath Tests

    func testHasTraversal() {
        XCTAssertTrue(SafePath.hasTraversal("../etc/passwd"))
        XCTAssertTrue(SafePath.hasTraversal("~/file"))
        XCTAssertFalse(SafePath.hasTraversal("normal/path"))
    }

    func testIsSafe() {
        XCTAssertTrue(SafePath.isSafe("safe/path"))
        XCTAssertFalse(SafePath.isSafe("../unsafe"))
    }

    func testSanitizeFilename() {
        XCTAssertEqual(SafePath.sanitizeFilename("file<>name"), "file__name")
        XCTAssertEqual(SafePath.sanitizeFilename("..secret"), "__secret")
    }

    func testSafeJoin() {
        XCTAssertEqual(SafePath.safeJoin(base: "/base", parts: ["a", "b"]), "/base/a/b")
        XCTAssertNil(SafePath.safeJoin(base: "/base", parts: ["../etc"]))
    }

    // MARK: - SafeEmail Tests

    func testIsValidEmail() {
        XCTAssertTrue(SafeEmail.isValid("user@example.com"))
        XCTAssertFalse(SafeEmail.isValid("not-an-email"))
        XCTAssertFalse(SafeEmail.isValid("@invalid.com"))
        XCTAssertFalse(SafeEmail.isValid("user@.com"))
    }

    func testSplitEmail() {
        let parts = SafeEmail.split("user@example.com")
        XCTAssertNotNil(parts)
        XCTAssertEqual(parts?.localPart, "user")
        XCTAssertEqual(parts?.domain, "example.com")
    }

    func testNormalizeEmail() {
        XCTAssertEqual(SafeEmail.normalize("User@EXAMPLE.COM"), "User@example.com")
    }

    // MARK: - SafeUrl Tests

    func testParseUrl() {
        let parsed = SafeUrl.parse("https://example.com:8080/path?query=1#frag")
        XCTAssertNotNil(parsed)
        XCTAssertEqual(parsed?.scheme, "https")
        XCTAssertEqual(parsed?.host, "example.com")
        XCTAssertEqual(parsed?.port, 8080)
        XCTAssertEqual(parsed?.path, "/path")
        XCTAssertEqual(parsed?.query, "query=1")
        XCTAssertEqual(parsed?.fragment, "frag")
    }

    func testIsValidUrl() {
        XCTAssertTrue(SafeUrl.isValid("https://example.com"))
        XCTAssertFalse(SafeUrl.isValid("not a url"))
    }

    func testIsHttps() {
        XCTAssertTrue(SafeUrl.isHttps("https://secure.com"))
        XCTAssertFalse(SafeUrl.isHttps("http://insecure.com"))
    }

    // MARK: - SafeNetwork Tests

    func testIsValidIPv4() {
        XCTAssertTrue(SafeNetwork.isValidIPv4("192.168.1.1"))
        XCTAssertFalse(SafeNetwork.isValidIPv4("invalid"))
        XCTAssertFalse(SafeNetwork.isValidIPv4("256.1.1.1"))
    }

    func testIsPrivate() {
        XCTAssertTrue(SafeNetwork.isPrivate("192.168.1.1"))
        XCTAssertTrue(SafeNetwork.isPrivate("10.0.0.1"))
        XCTAssertTrue(SafeNetwork.isPrivate("172.16.0.1"))
        XCTAssertFalse(SafeNetwork.isPrivate("8.8.8.8"))
    }

    func testIsLoopback() {
        XCTAssertTrue(SafeNetwork.isLoopback("127.0.0.1"))
        XCTAssertFalse(SafeNetwork.isLoopback("192.168.1.1"))
    }

    func testIsPublic() {
        XCTAssertTrue(SafeNetwork.isPublic("8.8.8.8"))
        XCTAssertFalse(SafeNetwork.isPublic("192.168.1.1"))
    }

    // MARK: - SafeCrypto Tests

    func testConstantTimeCompare() {
        XCTAssertTrue(SafeCrypto.constantTimeCompare("secret", "secret"))
        XCTAssertFalse(SafeCrypto.constantTimeCompare("secret", "other"))
        XCTAssertTrue(SafeCrypto.constantTimeCompare("", ""))
    }

    func testSecureZero() {
        var data: [UInt8] = [1, 2, 3, 4]
        SafeCrypto.secureZero(&data)
        XCTAssertTrue(data.allSatisfy { $0 == 0 })
    }
}

File: bindings/kotlin/src/test/kotlin/com/hyperpolymath/proven/ProvenTests.kt
// SPDX-License-Identifier: PMPL-1.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath

package com.hyperpolymath.proven

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlin.test.assertFalse
import kotlin.test.assertNotNull

class ProvenTests {

    // SafeMath Tests

    @Test
    fun testSafeDiv() {
        assertEquals(5, SafeMath.div(10, 2))
        assertNull(SafeMath.div(10, 0))
    }

    @Test
    fun testSafeMod() {
        assertEquals(1, SafeMath.mod(10, 3))
        assertNull(SafeMath.mod(10, 0))
    }

    @Test
    fun testSafeAdd() {
        assertEquals(3, SafeMath.add(1, 2))
        assertNull(SafeMath.add(Int.MAX_VALUE, 1))
    }

    @Test
    fun testSafeSub() {
        assertEquals(2, SafeMath.sub(5, 3))
        assertNull(SafeMath.sub(Int.MIN_VALUE, 1))
    }

    @Test
    fun testSafeMul() {
        assertEquals(12, SafeMath.mul(3, 4))
        assertNull(SafeMath.mul(Int.MAX_VALUE, 2))
    }

    // SafeString Tests

    @Test
    fun testEscapeHtml() {
        assertEquals("&lt;script&gt;", SafeString.escapeHtml("<script>"))
        assertEquals("a &amp; b", SafeString.escapeHtml("a & b"))
        assertEquals("&quot;quoted&quot;", SafeString.escapeHtml("\"quoted\""))
    }

    @Test
    fun testEscapeSql() {
        assertEquals("it''s", SafeString.escapeSql("it's"))
    }

    @Test
    fun testEscapeJs() {
        assertEquals("line\\nbreak", SafeString.escapeJs("line\nbreak"))
        assertEquals("tab\\there", SafeString.escapeJs("tab\there"))
    }

    @Test
    fun testTruncateSafe() {
        assertEquals("he...", SafeString.truncateSafe("hello world", 5))
        assertEquals("hi", SafeString.truncateSafe("hi", 10))
    }

    // SafePath Tests

    @Test
    fun testHasTraversal() {
        assertTrue(SafePath.hasTraversal("../etc/passwd"))
        assertTrue(SafePath.hasTraversal("~/file"))
        assertFalse(SafePath.hasTraversal("normal/path"))
    }

    @Test
    fun testIsSafe() {
        assertTrue(SafePath.isSafe("safe/path"))
        assertFalse(SafePath.isSafe("../unsafe"))
    }

    @Test
    fun testSanitizeFilename() {
        assertEquals("file__name", SafePath.sanitizeFilename("file<>name"))
        assertEquals("__secret", SafePath.sanitizeFilename("..secret"))
    }

    @Test
    fun testSafeJoin() {
        assertEquals("/base/a/b", SafePath.safeJoin("/base", listOf("a", "b")))
        assertNull(SafePath.safeJoin("/base", listOf("../etc")))
    }

    // SafeEmail Tests

    @Test
    fun testIsValidEmail() {
        assertTrue(SafeEmail.isValid("user@example.com"))
        assertFalse(SafeEmail.isValid("not-an-email"))
        assertFalse(SafeEmail.isValid("@invalid.com"))
        assertFalse(SafeEmail.isValid("user@.com"))
    }

    @Test
    fun testSplitEmail() {
        val parts = SafeEmail.split("user@example.com")
        assertNotNull(parts)
        assertEquals("user", parts.localPart)
        assertEquals("example.com", parts.domain)
    }

    @Test
    fun testNormalizeEmail() {
        assertEquals("User@example.com", SafeEmail.normalize("User@EXAMPLE.COM"))
    }

    // SafeUrl Tests

    @Test
    fun testParseUrl() {
        val parsed = SafeUrl.parse("https://example.com:8080/path?query=1#frag")
        assertNotNull(parsed)
        assertEquals("https", parsed.scheme)
        assertEquals("example.com", parsed.host)
        assertEquals(8080, parsed.port)
        assertEquals("/path", parsed.path)
        assertEquals("query=1", parsed.query)
        assertEquals("frag", parsed.fragment)
    }

    @Test
    fun testIsValidUrl() {
        assertTrue(SafeUrl.isValid("https://example.com"))
        assertFalse(SafeUrl.isValid("not a url"))
    }

    @Test
    fun testIsHttps() {
        assertTrue(SafeUrl.isHttps("https://secure.com"))
        assertFalse(SafeUrl.isHttps("http://insecure.com"))
    }

    // SafeNetwork Tests

    @Test
    fun testIsValidIPv4() {
        assertTrue(SafeNetwork.isValidIPv4("192.168.1.1"))
        assertFalse(SafeNetwork.isValidIPv4("invalid"))
        assertFalse(SafeNetwork.isValidIPv4("256.1.1.1"))
    }

    @Test
    fun testIsPrivate() {
        assertTrue(SafeNetwork.isPrivate("192.168.1.1"))
        assertTrue(SafeNetwork.isPrivate("10.0.0.1"))
        assertTrue(SafeNetwork.isPrivate("172.16.0.1"))
        assertFalse(SafeNetwork.isPrivate("8.8.8.8"))
    }

    @Test
    fun testIsLoopback() {
        assertTrue(SafeNetwork.isLoopback("127.0.0.1"))
        assertFalse(SafeNetwork.isLoopback("192.168.1.1"))
    }

    @Test
    fun testIsPublic() {
        assertTrue(SafeNetwork.isPublic("8.8.8.8"))
        assertFalse(SafeNetwork.isPublic("192.168.1.1"))
    }

    // SafeCrypto Tests

    @Test
    fun testConstantTimeCompare() {
        assertTrue(SafeCrypto.constantTimeCompare("secret", "secret"))
        assertFalse(SafeCrypto.constantTimeCompare("secret", "other"))
        assertTrue(SafeCrypto.constantTimeCompare("", ""))
    }

    @Test
    fun testSecureZero() {
        val data = byteArrayOf(1, 2, 3, 4)
        SafeCrypto.secureZero(data)
        assertTrue(data.all { it == 0.toByte() })
    }
}
